src/gc-arena/src/gc.rs

```diff
@@ -86,6 +87,48 @@ impl<'gc, T: 'gc + Collect> Gc<'gc, T> {
     }
 
     pub fn as_ptr(gc: Gc<'gc, T>) -> *const T {
-        unsafe { gc.ptr.as_ref().value.get() }
+        let ptr: *mut GcBox<T> = NonNull::as_ptr(gc.ptr);
+
+        // SAFETY: This cannot go through Deref::deref or create temporary references
+        // because stacked borrows requires that any new (temporary) shared ref
+        // that gets created from a raw pointer invalidates any previous
+        // tag associated with that pointer. So, this gets a pointer to the
+        // underlying value without creating any temporary references.
+        unsafe { UnsafeCell::raw_get(ptr::addr_of!((*ptr).value) as _) }
+    }
+
+    unsafe fn from_inner(ptr: NonNull<GcBox<T>>) -> Self {
+        Self {
+            ptr,
+            _invariant: PhantomData,
+        }
+    }
+
+    unsafe fn from_ptr(ptr: *mut GcBox<T>) -> Self {
+        unsafe { Self::from_inner(NonNull::new_unchecked(ptr)) }
+    }
+
+    /// Construct a [Gc] from a raw pointer
+    ///
+    /// # Safety
+    ///
+    /// The raw pointer passed in must come from a call to [Gc::as_ptr]
```

Comment:
This isnt sufficient - we also need th3ballocstion to still exist. User code generally can't know when collection is going to run, or whether an object is white, so I'm not sure what conditions we should impose here.

Reply:
Hmm, I see what you're saying.  Any code working with raw pointers across GC calls needs to preserve the live-ness of the object it points to.  I think it would be sufficient to require that pointers come from a type that traces this pointer (so it doesn't get collected) and be persisted in a type that trace it:

```rust
/// Contrived example, but you could imagine I'm stuffing a pointer into a usize or f64 via e.g. [nanval](https://crates.io/crates/nanval)
pub struct Value<'gc, T: Collect> {
    ptr: *const T,
    _phantom: Invariant<'gc>,
}

unsafe impl<'gc, T: Collect> Collect for Value<'gc, T> {
    fn trace(&self, cc: CollectionContext {
       // SAFETY: we're tracing this pointer to make sure it stays valid
       let gc = unsafe { Gc::from_raw(self.ptr) };
       cc.trace(gc);
    }
}

impl<'gc> Value<'gc, T> {
    pub fn as_gc(&self) -> Gc<'gc, T> {
        // SAFETY: we have a collect impl that traces this pointer, so it should remain valid as long as we hold it
        unsafe { Gc::from_raw(self.ptr) }
    }
}
```